src/networks/invertible_network_conditional_hint_multiscale.jl

```diff
@@ -100,41 +100,41 @@ end
 NetworkMultiScaleConditionalHINT3D(args...;kw...) = NetworkMultiScaleConditionalHINT(args...; kw..., ndims=3)
 
 # Concatenate states Zi and final output
-function cat_states(XY_save::AbstractArray{Array, 2}, X::AbstractArray{Float32, 4}, Y::AbstractArray{Float32, 4})
-    X_full = []
-    Y_full = []
+function cat_states(XY_save::Array{AbstractArray, 2}, X::AbstractArray{Float32, 4}, Y::AbstractArray{Float32, 4})
+    X_full = cuzeros(X,0)
+    Y_full = cuzeros(X,0)
 
     for j=1:size(XY_save, 1)
         X_full = cat(X_full, vec(XY_save[j, 1]); dims=1)
         Y_full = cat(Y_full, vec(XY_save[j, 2]); dims=1)
     end
-    X_full = cat(X_full, vec(X); dims=1) #NOTE THIS MAKES THINGS INTO A NORMAL non-cuda ARRAY. 
+    X_full = cat(X_full, vec(X); dims=1) 
     Y_full = cat(Y_full, vec(Y); dims=1)
 
-
-    return Float32.(X_full), Float32.(Y_full)  # convert to Array{Float32, 1}
+    return X_full, Y_full 
 end
 
 # Split 1D vector in latent space back to states Zi
-function split_states(XY_dims::AbstractArray{Tuple, 1}, X_full::AbstractArray{Float32, 1}, Y_full::AbstractArray{Float32, 1})
+function split_states(XY_dims::AbstractArray{Array, 1}, X_full::AbstractArray{Float32, 1}, Y_full::AbstractArray{Float32, 1})
     L = length(XY_dims) + 1
-    XY_save = Array{Array}(undef, L-1, 2)
+    XY_save = Array{AbstractArray}(undef, L-1, 2)
```

Comment:
You can't do that. `AbstractArray` as its name show is an abstract type, not an instantiated type. You need to initialize with an actual type like Array or Cuarray, so in this case something like `typeof(X_full)`

Reply:
Can't do that or shouldn't? This code runs. Doesnt the undef mean that it wont initialize?

Reply:
`undef` is fine, `AbstractArray` is not, it''s an abstract type not an implemented type, so it should only be used for dispatch not to create an object.